### ISSUE TITLE
`project` doesn't incorporate `tsconfigRootDir` or work together with it

### ISSUE THREAD

**Issue by dPowNextdoor** [typescript-eslint/typescript-eslint#4439]
## Gist

Only `parserOptions.project` prefills ESLint fields like `jsx`, `lib`, etc., ignoring the fact that more tsconfig.json files may be defined in `parserOptions.tsconfigRootDir`. In reality, all tsconfig.json files should be considered if they exist. Even if that were complicated to calculate, you could allow the `project` entry to define `jsx`, `lib`, etc. and then accept deviations based on non-root-level configs.

TL;DR Why can't we specify both `parserOptions.project` as the root-level tsconfig.json to prefill other ESLint fields from, *and* `parserOptions.tsconfigRootDir` to allow other directories, extensions, etc. to be included in linting?

Or, even better, load them all at once! Work like `tsc` does and parse all directories, extensions, libs, etc. based on where the tsconfig.json and its `rootDir`, `baseUrl`, `include`, `exclude`, etc. exist and parse them from there.

**Checks**

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```javascript
const tsconfig = require('./tsconfig.json');

module.exports = {
    // ... all the necessary stuff + rules + etc.
    parser: '@typescript-eslint/parser',
    parserOptions: {
        ecmaVersion: 'latest',
        sourceType: 'module',
        ecmaFeatures: {
            jsx: true,
        },
        // We can't specify `project: tsconfigPath` b/c that only allows
        // ONE SINGLE tsconfig.json file to be parsed, ignoring the others.
        // Since we have a tsconfig.json in both the root and tests/ directories,
        // specify the top-level dir from which to find all nested tsconfig files.
        tsconfigRootDir: rootDir,
        // Extra extensions other than `.[tj]sx?` to be defined explicitly
        extraFileExtensions: [ '.mjs', '.cjs' ],
        /************************
        ** IMPORTANT PART HERE **
        ************************/
        // Since we aren't setting `project`, we unfortunately have to set these even
        // though they're ALREADY SET (yes, it's a bug from the TS-ESLint-parser team).
        jsxPragma: null,
        lib: tsconfig.compilerOptions.lib,
    },
    // ... more rules, etc.
};
```

```TS
// TypeScript file example can be anything; doesn't matter b/c this is a parser issue
```

```javascript
// tsconfig.json
{
    "compilerOptions": {
        "target": "ESNext",
        "lib": [
            "ESNext",
            "DOM",
            "DOM.Iterable"
        ],
        "module": "ESNext",
        "moduleResolution": "node",
        "strict": true,
        "jsx": "react-jsx",
        "allowJs": true,
        "esModuleInterop": true,
        "allowSyntheticDefaultImports": true,
        "experimentalDecorators": true,
        "emitDecoratorMetadata": true,
        "resolveJsonModule": true,
        "forceConsistentCasingInFileNames": true,
        "newLine": "lf",
        "incremental": true,
        "isolatedModules": true,
        "outDir": "dist",
        "declaration": true,
        "declarationDir": "dist/types",
        "declarationMap": true,
        "baseUrl": ".",
        "paths": {
            "@/*": [
                "src/*"
            ],
            "/*": [
                "*"
            ]
        }
    },
    "include": [
        "./src/**/*"
    ],
    "exclude": [
        "./node_modules"
    ],
    "ts-node": {
        "require": [
            "tsconfig-paths/register"
        ],
        "preferTsExts": true,
        "experimentalReplAwait": true,
        "transpileModule": false,
        "compilerOptions": {
            "module": "CommonJS",
            "isolatedModules": false
        }
    }
}
```

```javascript
// tests/tsconfig.json - only needed for import aliases autocompletion in IDE
{
    "extends": "../tsconfig",
    "include": [
        "../src",
        "../tests"
    ]
}

```

**Expected Result**

I expect that `parserOptions.project` not be the only field to offer default auto-filling of linting fields.

**Actual Result**

Even though:

1. There is a root-level tsconfig.json which specifies `jsx`, `lib`, and all other possible fields that might be relevant for linting.
2. The tsconfig.json files below that root-level one *extends* it.

I still must specify these ESLint fields manually, duplicating code in many places.

**Additional Info**

ESLint threw an error that `These (file extensions|directories) aren't specified in your "includes" field, please add them`. Yet, if I were able to specify both `project` and `tsconfigRootDir`, this wouldn't happen!

It's extremely annoying (and very surprising) that I have to fill in these fields manually. Although `tsc` and even the incredibly unsupported `ts-node`/`tsconfig-paths` packages supported my multi-layer tsconfig.json structure, `@typescript-eslint/parser` somehow can't understand that there is a definition for `jsx`, `lib`, etc. that the other `tests/tsconfig.json` extends from.

Plus, I saw a performance decrease between the two. Probably b/c `project` is just guessing at what to do and `tsconfigRootDir` iterates through everything for each file. Realistically, specifying one could auto-load the configs for everything so options don't have to be re-calculated at each linting step.

**Versions**

| package                     | version   |
| --------------------------- | -------   |
| `@typescript-eslint/parser` | `5.9.1`   |
| `TypeScript`                | `4.5.4`   |
| `ESLint`                    | `7.32.0`  |
| `node`                      | `16.13.1` |

**Comment by bradzacher** [comment]
I'm not sure I understand what you're asking here.

`tsconfigRootDir` does nothing on its own.
It only serves to tell us the root of your project folder. Why? Because we if you provide relative paths, or relative globs to `parserOptions.project`, then we need to resolve the relativity against something.
Unfortunately ESLint doesn't tell us the path of the config file (for various completely valid reasons), so `tsconfigRootDir` is the mechanism through which we can orient ourselves on disk.
Without it, we can only orient ourselves using the current working directory - which is not constant - meaning the meaning of the relative path/glob isn't constant either (which is **bad**!).

When you provide `parserOptions.project` it tells our tooling to parse the project in type-aware mode. This allows type-aware lint rules to work, but does incur a performance cost due to the requirement to typecheck your project.

You can read more about type-aware linting in our docs:
- https://typescript-eslint.io/docs/linting/type-linting
- https://typescript-eslint.io/docs/linting/monorepo

And you can find the list of type-aware lint rules by looking for rules marked with 💭 in the rules list: https://typescript-eslint.io/rules/#supported-rules

---

If you don't turn on type-aware linting, we don't read your tsconfigs (or rather, we don't ask TS to typecheck the projects associated with your tsconfigs). There are many reasons for this, but it is on purpose. The duplicated config required is ultimately minimal (3 config fields I believe) which only needs to be done once. By making the user do this we save A LOT of complicated code which would be required to find, load, parse, and contextually match each non-type-aware tsconfig up with the relevant files it applies to.

**Comment by dPowNextdoor** [comment]
I probably threw the pieces of the puzzle down without explaining how they fit together very well, so I'll update the original post to clarify what I meant.

The issue wasn't so much a "oh no it's not automatic and I'm lazy" (although I admit that it sounded like that), it was that configuring it is a "one or the other" type of deal.

1. If I only put `project`, then `tests/`/`.test.ts` don't get linted and I get the same errors as (3).
2. If I only put `tsconfigRootDir`, then the fields from the root tsconfig don't get picked up/ESLint can't make assumptions about the code, meaning:
    - I have to duplicate code; not the biggest deal, but also not DRY so it's easy to forget to update ESLint when making a tsconfig change (I've worked pretty hard to make it the single source of truth even with other, non-TS code).
    - More importantly, it makes ESLint run slower; I can't say why, but I noticed up to a 20% increase in length of ESLint runtime (obv this varies, but it was always slower when I ran it this way).
3. If I put both `project` and `tsconfigRootDir`, then I get errors that 
```
0:0  error  Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: tests/MyFile.test.ts.
The file must be included in at least one of the projects provided
```

So it actually seems I'm not allowed to define both. I think I tried (3) first, got confused why it was breaking, then tried the other two methods; I posted after trying each of them, which inaccurately focused on the frustration of having to use `project` + duplicate code, when in reality it was that (a) both fields can't work together, and (b) the only one that does work is kind of the worst of all worlds (slower + not DRY).

**Comment by bradzacher** [comment]
> If I only put tsconfigRootDir, then the fields from the root tsconfig don't get picked up/ESLint can't make assumptions about the code

To clarify - as mentioned in my other comment:

`tsconfifRootDir` does nothing on its own because all it does is specifies the root ***folder*** that we should use to resolve relative paths.

`project` can be a string, or an array of strings. The value(s) can either be a path or a glob. The value(s) can either be absolute or relative.
If given a relative glob/path, then we will resolve the path relative to the current working directory. We cannot resolve the path relative to the config file because ESLint does not tell us the config file location for various reasons.

Because this behaviour can make configs dynamic and unpredictable, `tsconfigRootDir` exists to allow you to override the CWD resolution with a static path. If you provide `tsconfigRootDir`, then we resolve the relative paths passed in `project` relative to that directory, instead of CWD.

Relevant code:
https://github.com/typescript-eslint/typescript-eslint/blob/9eb0a5b3d760d467ae43e6281cace9bee6ca54e9/packages/typescript-estree/src/create-program/shared.ts#L65-L69

---

The config you provided in the issue only shows `tsconfigRootDir: rootDir`. Is `rootDir` a path to a tsconfig or folder? It should be the latter. Providing the former will cause incorrect behaviour as it is incorrect.

Our general recommendation is to set `tsconfigRootDir: __dirname` and define all your paths relative to your ESLint config.

For example, our config in this repo is:
https://github.com/typescript-eslint/typescript-eslint/blob/9eb0a5b3d760d467ae43e6281cace9bee6ca54e9/.eslintrc.js#L21-L29

---

> More importantly, it makes ESLint run slower; I can't say why, but I noticed up to a 20% increase in length of ESLint runtime

I can pretty much guarantee that this isn't due to specifying `tsconfigRootDir`. It's just a path to a folder - it doesn't change any internal logic. If you don't provide `project` it's only used exactly once here:

https://github.com/typescript-eslint/typescript-eslint/blob/9eb0a5b3d760d467ae43e6281cace9bee6ca54e9/packages/typescript-estree/src/parser.ts#L278-L279

If you can create a repro which shows that turning on just `tsconfigRootDir` slows down - I will gladly investigate further.

As I mentioned in my other comment, type-aware linting will slow down your lint run, but that requires `parserOptions.project` to be specified.

---

> (1) If I only put project, then tests//.test.ts don't get linted and I get the same errors as (3).
>
> (3) If I put both project and tsconfigRootDir, then I get errors that...

[This document (mentioned in my other comment) explains how to fix the issue](https://typescript-eslint.io/docs/linting/type-linting/#i-get-errors-telling-me-the-file-must-be-included-in-at-least-one-of-the-projects-provided)
